### PR TITLE
Gemma4: fix failed test cases

### PR DESCRIPTION
**Pull request by kaixuanliu**
# What does this PR do?

This PR did several things:
1. Skip some test cases that are not suitbale for gemma4 model
2. Fix bug when `attention_mask` is None(tests/models/gemma4/test_modeling_gemma4.py::Gemma4Audio2TextModelTest::test_eager_matches_fa2_generate)
3. fix some failed test cases related to `test_flash_attn_x_from_config`
4. Add XPU related Expectations

Fixes # (issue)

## Code Agent Policy

- [X] I confirm that this is not a pure code agent PR.

## Who can review?
@ydshieh pls help review

### CODE REVIEW

File: src/transformers/models/gemma4/modeling_gemma4.py [hunk 1]
```diff
@@ -1941,7 +1941,8 @@ def forward(
                 (self.config.attention_context_left - 1, self.config.attention_context_right)
             ),
         )
-        attention_mask = self._convert_4d_mask_to_blocked_5d(attention_mask)
+        if attention_mask is not None:
+            attention_mask = self._convert_4d_mask_to_blocked_5d(attention_mask)
```
Comment (ydshieh) [inline src/transformers/models/gemma4/modeling_gemma4.py:1945]: @Cyrilvallez  any opinion.

From PR descriptioin

> Fix bug when attention_mask is None(tests/models/gemma4/test_modeling_gemma4.py::Gemma4Audio2TextModelTest::test_eager_matches_fa2_generate)

File: tests/models/gemma4/test_modeling_gemma4.py [hunk 2]
```diff
@@ -126,6 +132,20 @@ def test_tp_generation_quantized(self):
...
+    )
+    def test_flash_attn_2_can_compile_with_attention_mask_None_without_graph_break(self):
+        pass
+
+    @unittest.skip(
+        "Under non-bf16 dtypes, MoE grouped_mm falls back to "
+        "_grouped_mm_fallback_backward which is incompatible with torch.compile."
+    )
+    def test_torch_compile_for_training(self):
+        pass
```
Comment (ydshieh) [inline tests/models/gemma4/test_modeling_gemma4.py:142]: OK for me. Just let one of @Cyrilvallez or @vasqu to also valid or comment

Comment (ydshieh) [reply tests/models/gemma4/test_modeling_gemma4.py:142]: They are indeed failing on our CI too

File: tests/models/gemma4/test_modeling_gemma4.py [hunk 3]
```diff
@@ -441,6 +461,54 @@ def test_num_layers_is_small(self):
...
+        # Gemma4 requires mm_token_type_ids in train mode, so we test in eval mode
+        self.flash_attn_from_config(attn_implementation="flash_attention_2", test_fwd_in_train=False)
+
+    @require_flash_attn_3
+    @require_torch_gpu
+    @mark.flash_attn_3_test
+    @slow
+    def test_flash_attn_3_from_config(self):
+        # Gemma4 requires mm_token_type_ids in train mode, so we test in eval mode
+        self.flash_attn_from_config(attn_implementation="flash_attention_3", test_fwd_in_train=False)
```
Comment (ydshieh) [inline tests/models/gemma4/test_modeling_gemma4.py]: @kaixuanliu I didn't see these 2 failing on our Flash Attn CI job.

Could you share more info / error logs ?

Comment (vasqu) [reply tests/models/gemma4/test_modeling_gemma4.py]: Our flash attn ci doesn have FA3 - I think it's hard to install because you need to compile from source and it's much longer than FA2 build from source

Maybe we could add a separate FA4 CI - not sure how stable it is tho since it's still in beta

File: tests/models/gemma4/test_modeling_gemma4.py (hunk 2 above)
Comment (vasqu) [reply tests/models/gemma4/test_modeling_gemma4.py:142]: Hmm, iirc the fallback should be compile compatible cc @IlyasMoutawwakil

File: tests/models/gemma4/test_modeling_gemma4.py [hunk 4]
```diff
@@ -441,6 +461,54 @@ def test_num_layers_is_small(self):
...
+    @unittest.skip("The base test does not pass image_position_ids and mm_token_type_ids required by Gemma4")
+    def test_flash_attn_3_inference_equivalence_right_padding(self):
+        pass
+
+    @unittest.skip("The base test does not pass image_position_ids and mm_token_type_ids required by Gemma4")
+    def test_flash_attn_4_inference_equivalence(self):
+        pass
+
+    @unittest.skip("The base test does not pass image_position_ids and mm_token_type_ids required by Gemma4")
+    def test_flash_attn_4_inference_equivalence_right_padding(self):
```
Comment (vasqu) [inline tests/models/gemma4/test_modeling_gemma4.py]: Can we have something like https://github.com/huggingface/transformers/blob/727741f533b1f47120db47ef5934296d07214d7e/tests/models/dia/test_modeling_dia.py#L258-L271

But for FA? Imo it will always be quite a lot to skip these manually like that

File: tests/models/gemma4/test_modeling_gemma4.py (hunk 2 above)
Comment (vasqu) [reply tests/models/gemma4/test_modeling_gemma4.py:142]: Which torch version is CI now at btw?

Comment (IlyasMoutawwakil) [reply tests/models/gemma4/test_modeling_gemma4.py:142]: it is torch compileable, just not any mode that uses cuda graphs (like max-autotune), where the torch.grouped_mm also fails on <sm90 (it also uses the fallback path)

File: tests/models/gemma4/test_modeling_gemma4.py (hunk 3 above)
Comment (kaixuanliu) [reply tests/models/gemma4/test_modeling_gemma4.py]: Well, for FA3 and FA4, on my env they are skipped as well. I can delete these two.

Comment (vasqu) [reply tests/models/gemma4/test_modeling_gemma4.py]: Ah no, see my comment below https://github.com/huggingface/transformers/pull/45568#discussion_r3161170061

Comment (ydshieh) [reply tests/models/gemma4/test_modeling_gemma4.py]: No, I mean for 

test_flash_attn_2_from_config

our CI is `[PASSED]`. So I am not sure why we need this fix, at least for FA2.

Our CI runner don't have FA3 or FA4, so they are skipped. But the question may still valid: do we really this fix?

Comment (kaixuanliu) [reply tests/models/gemma4/test_modeling_gemma4.py]: well, the code is added before this PR is merged: https://github.com/huggingface/transformers/pull/45454, it will crash for this case before [here](https://github.com/huggingface/transformers/pull/45454/changes#diff-663099e7a4d592c4a9bcad8f04f555f71d0475b8fa9e7ec399845d29ef5ba18dL2100) is removed. After this PR this case can pass. I will update the code.

File: tests/models/gemma4/test_modeling_gemma4.py (hunk 4 above)
Comment (kaixuanliu) [reply tests/models/gemma4/test_modeling_gemma4.py]: Thx for the advice. Is it OK now?

File: tests/models/gemma4/test_modeling_gemma4.py (hunk 2 above)
Comment (vasqu) [reply tests/models/gemma4/test_modeling_gemma4.py:142]: So it shouldn't fail iiuc?

File: tests/models/gemma4/test_modeling_gemma4.py (hunk 4 above)
Comment (vasqu) [reply tests/models/gemma4/test_modeling_gemma4.py]: Yep much better thanks

File: tests/models/gemma4/test_modeling_gemma4.py (hunk 2 above)
Comment (kaixuanliu) [reply tests/models/gemma4/test_modeling_gemma4.py:142]: In the skipped test case, they use mode="reduce-overhead", which use cuda graphs and will fail at [L238](https://github.com/huggingface/transformers/blob/main/src/transformers/integrations/moe.py#L238) with error: 
```
E                       torch._dynamo.exc.BackendCompilerFailed: backend='inductor' raised:
E                       GuardOnDataDependentSymNode: Could not guard on data-dependent expression Eq(u0, 0) (unhinted: Eq(u0, 0)).  (Size-like symbols: none)
E
E                       consider using data-dependent friendly APIs such as guard_or_false, guard_or_true and statically_known_true.
E                       Caused by: (oot/upstream/transformers/src/transformers/integrations/moe.py:238 in _grouped_mm_fallback_backward)
E                       For more information, run with TORCH_LOGS="dynamic"
E                       For extended logs when we create symbols, also add TORCHDYNAMO_EXTENDED_DEBUG_CREATE_SYMBOL="u0"
E                       If you suspect the guard was triggered from C++, add TORCHDYNAMO_EXTENDED_DEBUG_CPP=1
E                       For more debugging help, see https://docs.google.com/document/d/1HSuTTVvYH1pTew89Rtpeu84Ht3nQEFTYhAX3Ypa_xJs/edit?usp=sharing
E
E                       For C++ stack trace, run with TORCHDYNAMO_EXTENDED_DEBUG_CPP=1
E
E                       Set TORCHDYNAMO_VERBOSE=1 for the internal stack trace (please do this especially if you're reporting a bug to PyTorch). For even more developer context, set TORCH_LOGS="+dynamo"

/opt/venv/lib/python3.12/site-packages/torch/fx/experimental/symbolic_shapes.py:7862: BackendCompilerFailed
```

**Comment by Qodo-Free-For-OSS** [comment]
Hi, Integration tests add device-specific Expectations entries for XPU without a default fallback, so running these tests on an unsupported accelerator type (or an XPU generation not covered) can select an unintended expectation or raise if no expectation matches. This makes the tests more brittle to new device properties.

**Severity:** informational | **Category:** reliability

**How to fix:** Add default expectation fallback

**Agent prompt to fix** - you can give this to your LLM of choice:

> ### Issue description
> XPU expectations were added without a `(None, None)` default. This can make the test brittle when run on different XPU generations or unexpected device properties.
> 
> ### Fix Focus Areas
> - tests/models/gemma4/test_modeling_gemma4.py[534-542]
> - tests/models/gemma4/test_modeling_gemma4.py[575-593]
> - tests/models/gemma4/test_modeling_gemma4.py[621-629]
> - tests/models/gemma4/test_modeling_gemma4.py[675-681]
> - tests/models/gemma4/test_modeling_gemma4.py[745-755]
> 
> ### Recommended changes
> - Add a default expectation `(None, None): <existing cuda expectation>` if you want to preserve previous behavior for other devices.
> - Or, add additional XPU keys if multiple gens are expected to run these tests.
> - Ensure the intended device coverage is explicit to avoid accidental matching on future hardware.

---
*Found by [Qodo](https://qodo.ai) code review*

File: tests/models/gemma4/test_modeling_gemma4.py (hunk 2 above)
Comment (vasqu) [reply tests/models/gemma4/test_modeling_gemma4.py:142]: Can we adjust the comment then at least?

Comment (kaixuanliu) [reply tests/models/gemma4/test_modeling_gemma4.py:142]: Done.

File: tests/models/gemma4/test_modeling_gemma4.py (hunk 3 above)
Comment (vasqu) [reply tests/models/gemma4/test_modeling_gemma4.py]: cc @zucchini-nlp for viz, don't think it's super important but would still be nice to fix at some point ig

Comment (zucchini-nlp) [reply tests/models/gemma4/test_modeling_gemma4.py]: ` Gemma4 requires mm_token_type_ids in train mode, so we test in eval mode` > was fixed already no?

If not, @kaixuanliu can you open an issue and ping me there. I might forget to come back when bugs are reported under PRs 😅

Comment (kaixuanliu) [reply tests/models/gemma4/test_modeling_gemma4.py]: It's already fixed. And I have removed this part.

**Review by vasqu — APPROVED** [review]
It kind of got lost, sorry let me run slow tests just for sanity checking and then merge

**Comment by vasqu** [comment]
run-slow: gemma4
